design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,143 @@
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a multus userspace CNI is being used, for instance userspace cni from intel.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane and userspace cni with multus support.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. An EmptyDIR volume (shared-dir) and DownwardAPI (podinfo) volume are mounted on the virtlauncher pod to have the support for new interface.
+- These mounts will allow the virt-launcher pod to create a VM with an additional interface which can be reached using the vhostSocket mentioned in the DownwardAPI.
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. **Once the VM spec virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) "/var/lib/cni/usrcni/" is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane. This socket file acts as an UNIX socket between the VM interface and the dataplane running in the UserSpace.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
+    2. **The CNI will mount the shared-dir i.e., EmptyDir volume's host path /var/lib/kubelet/pods/<podID>/volumes/kubernetes.io~empty-dir/shared-dir . This will be deleted, to delete the vHostUser Interface while deleteing the VM and/or virt-launcher pod.**
+    3. **The CNI will update the virt-launcher pod annotations with vhostsocket-file name and details.**
+    4. **The virt-launcher reads the DownwardAPI volume and retrieves the vhostsocket-file name specified in pod annotations and uses it to create a unix socket in the VM, while launching the VM using libvirtd.**
+    4. **The virt-launcher is modified to skip establishing the networking between VM and the virt-launcher pod using bridge(Refer in Kubevirt Networking section in https://kubevirt.io/2018/KubeVirt-Network-Deep-Dive.html).**
+    5. **Instead of using the bridge through the launcher pod, the vHostUser interface of the VM will be directly connected to the DPDK datplane using the vhost socket**
```

Comment:
It sounds like `<source type='unix' path='/tmp/vhost1.sock' mode='server'/>` will be used so QEMU creates the UNIX domain socket and then something must trigger DPDK to connect to this socket?

Which component sends a command to DPDK to establish the UNIX domain socket connection that QEMU is waiting for so the VM can start?

Reply:
@stefanha , I didn't get your question.

Reply:
The vhost-user protocol allows the frontend and backend to act as the client or server. This is configurable and QEMU supports both acting as a socket client and a server.

In your proposal the socket seems to be created by QEMU (the frontend) and DPDK (the backend) connects to the UNIX domain socket. (I say this because you wrote "The __virt-launcher__ reads the DownwardAPI volume and retrieves the vhostsocket-file name specified in pod annotations and uses it to __create__ a unix socket in the VM".)

My question is about the order in which QEMU creates the UNIX domain socket and DPDK connects to it:

How does DPDK know when it is time to connect to the UNIX domain socket since the socket is created after virt-launcher Pod launches?

Does DPDK need to attempt to reconnect in a loop or is there a step where DPDK is notified that it's time to connect?

Reply:
@stefanha , The UNIX socket will be created in the host in the custom path linked to the shared directory  and the qemu then connects to it. The socket details will be passed to the qemu by the virtualization-launcher while creating the VM, which gets this information from the downwardAPI passed by the Kubemanager.

Reply:
Does the NAD below need to be updated too?

```
        "host": {
                "engine": "ovs-dpdk",
                "iftype": "vhostuser",
                "netType": "bridge",
                "vhost": {
                        "mode": "client"
                },
                "bridge": {
                        "bridgeName": "br-dpdk0"
                }
        },
...
        "container": {
                "engine": "ovs-dpdk",
                "iftype": "vhostuser",
                "netType": "interface",
                "vhost": {
                        "mode": "server"
                }
        },
```

Maybe "mode" needs to be reversed for "host" and "container".

Do you intend to support both configurations, or just DPDK acting as the server and QEMU acting as the client?